granulate_utils/config_feeder/client/yarn/collector.py

```diff
@@ -0,0 +1,63 @@
+import logging
+from typing import Any, Dict, Optional, Union
+
+from requests.exceptions import ConnectionError
+
+from granulate_utils.config_feeder.client.base import ConfigCollectorBase
+from granulate_utils.config_feeder.client.yarn.models import YarnConfig
+from granulate_utils.config_feeder.client.yarn.utils import (
+    RM_DEFAULT_ADDRESS,
+    WORKER_ADDRESS,
+    detect_resource_manager_address,
+    get_yarn_properties,
+)
+from granulate_utils.config_feeder.core.models.node import NodeInfo
+
+
+class YarnConfigCollector(ConfigCollectorBase):
+    def __init__(self, *, max_retries: int = 20, logger: Union[logging.Logger, logging.LoggerAdapter] = None) -> None:
+        super().__init__(max_retries=max_retries, logger=logger)
+        self._resource_manager_address = RM_DEFAULT_ADDRESS
+        self._is_address_detected = False
+
+    async def collect(self, node_info: NodeInfo) -> Optional[YarnConfig]:
+        if config := await (self._get_master_config() if node_info.is_master else self._get_worker_config()):
+            return YarnConfig(
+                config=config,
+            )
+
+        self.logger.error("failed to collect node config", extra=node_info.__dict__)
+        return None
+
+    async def rm_request(self, path: str) -> Optional[Dict[str, Any]]:
+        result: Optional[Dict[str, Any]] = None
+        try:
+            result = await self._fetch(f"{self._resource_manager_address}{path}")
+        except ConnectionError:
+            if self._is_address_detected:
+                self.logger.error(f"could not connect to {self._resource_manager_address}")
+                return None
+            self.logger.warning(f"ResourceManager not found at {self._resource_manager_address}")
+            if address := await detect_resource_manager_address(logger=self.logger):
+                self._is_address_detected = True
+                self._resource_manager_address = address
+                self.logger.debug(f"found ResourceManager address: {address}")
+                result = await self._fetch(f"{address}{path}")
+            else:
+                self.logger.error("could not resolve ResourceManager address")
+        return result
+
+    async def node_request(self, path: str) -> Optional[Dict[str, Any]]:
+        try:
+            return await self._fetch(f"{WORKER_ADDRESS}{path}")
+        except ConnectionError:
+            self.logger.warning(f"failed to connect to {WORKER_ADDRESS}")
+        return None
+
+    async def _get_master_config(self) -> Optional[Dict[str, Any]]:
+        config: Optional[Dict[str, Any]] = await self.rm_request("/conf")
+        return get_yarn_properties(config) if config else None
+
+    async def _get_worker_config(self) -> Optional[Dict[str, Any]]:
+        config: Optional[Dict[str, Any]] = await self.node_request("/conf")
```

Comment:
Did you check supported versions of this endpoint? I couldn't find it in the docs. Also, does it report the original config from disk, or the most recent config? Please add to the method's documentation.

Reply:
this feature is very old https://issues.apache.org/jira/browse/HADOOP-6408
minimal required version is 2.9 because of https://issues.apache.org/jira/browse/HADOOP-13588

Reply:
I'm pretty sure we have clients with Hadoop < 2.9, so we might have to implement a fallback in the future, but for now it's fine. Please handle failure to resolve it with a meaningful error log and also document the version constraints, so that we'll immediately know what the issue is when we encounter a cluster with Hadoop < 2.9.

Reply:
@slicklash is the current logging at the top of the module is enough?

Reply:
yes